Kitodo/src/main/java/de/sub/goobi/helper/ProjectHelper.java

```diff
@@ -39,157 +34,92 @@
-     *         separate classes or to create a new class with these properties
+     *         according to the definition that only steps shared by all processes
+     *         are returned. The workflow order is returned according to the average
+     *         order return by a grouping by step title consider workflow structure
+     *         to be a prototype, it would probably make things easier, to either
+     *         assemble the underlying construction in separate classes or to create
+     *         a new class with these properties
      */
-
     @SuppressWarnings("unchecked")
-    public static synchronized List<StepInformation> getProjectWorkFlowOverview(Project project) {
-        Long totalNumberOfProc = 0l;
-        Long totalNumberOfImages = 0l;
-
-        Session session = Helper.getHibernateSession();
-
-        Criteria critTotals = session.createCriteria(Process.class, "proc");
-        critTotals.add(Restrictions.eq("proc.template", Boolean.FALSE));
-        critTotals.add(Restrictions.eq("proc.project", project));
-
-        ProjectionList proList = Projections.projectionList();
-
-        proList.add(Projections.count("proc.id"));
-        proList.add(Projections.sum("proc.sortHelperImages"));
-
-        critTotals.setProjection(proList);
-
-        List<Object> list = critTotals.list();
-
-        for (Object obj : list) {
-            Object[] row = (Object[]) obj;
-
-            totalNumberOfProc = (Long) row[FieldList.totalProcessCount.fieldLocation];
-            totalNumberOfImages = (Long) row[FieldList.totalImageCount.fieldLocation];
-            ;
-        }
-
-        proList = null;
-        list = null;
-
-        Criteria critSteps = session.createCriteria(Task.class);
-
-        critSteps.createCriteria("process", "proc");
-        critSteps.addOrder(Order.asc("ordering"));
-
-        critSteps.add(Restrictions.eq("proc.template", Boolean.FALSE));
-        critSteps.add(Restrictions.eq("proc.project", project));
+    public static synchronized List<StepInformation> getProjectWorkFlowOverview(Project project) throws DataException {
+        ServiceManager serviceManager = new ServiceManager();
 
-        proList = Projections.projectionList();
+        Long totalNumberOfProcesses = serviceManager.getProcessService()
+                .findNumberOfNotTemplateProcessesForProjectId(project.getId());
+        Double totalNumberOfImages = serviceManager.getProcessService()
+                .findAmountOfImagesForNotTemplatesAndProjectId(project.getId());
 
-        proList.add(Projections.groupProperty(("title")));
-        proList.add(Projections.count("id"));
-        proList.add(Projections.avg("ordering"));
-
-        critSteps.setProjection(proList);
+        List<Task> tasks = serviceManager.getTaskService()
+                .getTasksForProjectHelper(project.getId());
+        List<Long> tasksSize = serviceManager.getTaskService()
```

Comment:
The amount of `tasks` and the number in `taskSize` is only equal if there no data inserted between both calls. Value of `taskSize` should be gathered from  `tasks.size()` call instead of quering the database again.

Reply:
IT is not exactly the same value. It counts ids of tasks inside Group By clausule.